### PR TITLE
perf: small improvement to TensorProduct

### PR DESCRIPTION
**Pull request by ChrisHughes24**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by ChrisHughes24** [comment]
!bench

**Comment by riccardobrasca** [comment]
@mattrobball there is a bunch of others PRs by Chris similar to this one. You can maybe have a look, thanks!

File: Mathlib/LinearAlgebra/TensorProduct.lean [hunk 1]
```diff
@@ -94,12 +94,16 @@ def addMonoid : AddMonoid (M ⊗[R] N) :=
 protected instance add : Add (M ⊗[R] N) :=
   (addConGen (TensorProduct.Eqv R M N)).hasAdd
 
+instance zero : Zero (M ⊗[R] N) :=
+  (addConGen (TensorProduct.Eqv R M N)).addMonoid.toZero
+
 instance addZeroClass : AddZeroClass (M ⊗[R] N) :=
   { (addConGen (TensorProduct.Eqv R M N)).addMonoid with
```
Comment (mattrobball) [inline Mathlib/LinearAlgebra/TensorProduct.lean:101]: Why is this not `.addZeroClass` to start with?

File: Mathlib/LinearAlgebra/TensorProduct.lean [hunk 2]
```diff
@@ -94,12 +94,16 @@ def addMonoid : AddMonoid (M ⊗[R] N) :=
...
     /- The `toAdd` field is given explicitly as `TensorProduct.add` for performance reasons.
     This avoids any need to unfold `Con.addMonoid` when the type checker is checking
     that instance diagrams commute -/
-    toAdd := TensorProduct.add _ _ }
+    toAdd := TensorProduct.add _ _
+    toZero := TensorProduct.zero _ _ }
 
 instance addSemigroup : AddSemigroup (M ⊗[R] N) :=
   { (addConGen (TensorProduct.Eqv R M N)).addMonoid with
```
Comment (mattrobball) [inline Mathlib/LinearAlgebra/TensorProduct.lean:109]: Similarly here.

**Review by mattrobball — COMMENTED** [review]
Using `with` and `.addMonoid` seems problematic to build classes which are parents of `AddMonoid`, would changing to match the exact instances from `GroupTheory.Congruence` do just as well?

File: Mathlib/LinearAlgebra/TensorProduct.lean (hunk 1 above)
Comment (ChrisHughes24) [reply Mathlib/LinearAlgebra/TensorProduct.lean:101]: There's no `addZeroClass` instance on `AddCon`.

Comment (ChrisHughes24) [reply Mathlib/LinearAlgebra/TensorProduct.lean:101]: Sorry, there is, but the file is in a funny order so I didn't see it.

**Comment by ChrisHughes24** [comment]
!bench

**Comment by ChrisHughes24** [comment]
> Here are the [benchmark results](http://speed.lean-fro.org/mathlib4/run-detail/94617e10-44bc-4aaf-934e-eb5214362b67) for commit [30f53bd](https://github.com/leanprover-community/mathlib4/commit/30f53bd7bd8d5b886c8bd4611ca3a8023202d272). There were [significant changes](http://speed.lean-fro.org/mathlib4/compare/23eda3c9-f1e0-46f8-a356-398e717b2226/to/94617e10-44bc-4aaf-934e-eb5214362b67) against commit [3d72934](https://github.com/leanprover-community/mathlib4/commit/3d729341e47389b74ece4d01d6cf1a53c7526c82):
> 
> ```diff
>   Benchmark                                            Metric         Change
>   ==========================================================================
> + ~Mathlib.Algebra.Category.ModuleCat.Monoidal.Basic   instructions    -6.3%
> + ~Mathlib.LinearAlgebra.FreeModule.PID                instructions   -20.8%
> ```

Very small slowdown in total instructions.

**Comment by mattrobball** [comment]
That is probably noise. I should have been more clear. I wanted to know if not going through `AddMonoid` and instead just applying `.addZeroClass` `.addSemigroup` etc. directly would have the same effect. 

Does avoiding unifying `nsmul` account for all the speedup? If not, this kind of expansion goes against the intuition I have generally from things like leanprover/lean4#2478 and I wonder if other changes elsewhere that make terms more compact and remove data from their construction would improve this file.

**Comment by ChrisHughes24** [comment]
I don't understand what `nsmul` has to do with this PR. I think the only benefits probably come from the changes to `Zero`. I also don't know what you mean by "applying `.addZeroClass` directly"

**Comment by mattrobball** [comment]
Aren't the fields populating with `AddMonoid.x` which requires checking all the inputs of `AddMonoid` during unification?

I mean `instance : AddZeroClass Y := X.addZeroClass` without breaking out the `Add` and `Zero`

**Pull request closed without merging**